File: client-mobile/app/src/main/java/com/example/polyquiz/match/domain/AnswerService.kt
package com.example.polyquiz.match.domain

import android.os.Handler
import android.os.Looper
import android.util.Log
import com.example.polyquiz.constants.AnswerCorrectness
import com.example.polyquiz.constants.AnswerEvents
import com.example.polyquiz.constants.ChoiceInfo
import com.example.polyquiz.constants.Feedback
import com.example.polyquiz.constants.GradesInfo
import com.example.polyquiz.constants.LongAnswerInfo
import com.example.polyquiz.constants.MatchEvents
import com.example.polyquiz.constants.UserInfo
import com.example.vanillaprototype.socket.SocketHandler
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.setValue
import com.example.polyquiz.auth.domain.AuthViewModel
import com.example.polyquiz.constants.MatchContext
import org.json.JSONObject
import java.util.Timer
import kotlin.concurrent.schedule

object AnswerService {
    var playersAnswers by mutableStateOf<List<LongAnswerInfo>>(emptyList())
    var feedback by mutableStateOf(Feedback(0, AnswerCorrectness.WRONG, emptyList()))
    var gradeAnswers by mutableStateOf(false)
    var isGradingComplete by mutableStateOf(false)
    var showFeedback by mutableStateOf(false)
    var isNextQuestionButtonEnabled by mutableStateOf(false)
    var isSelectionEnabled by mutableStateOf(false)
    var correctAnswer by mutableStateOf<List<String>>(emptyList())
    var answerCorrectness by mutableStateOf(AnswerCorrectness.WRONG)
    var playerScore by mutableStateOf(0)
    var bonusPoints by mutableStateOf(0)
    var isTimesUp by mutableStateOf(false)
    var isEndGame by mutableStateOf(false)
    var currentLongAnswer by mutableStateOf("")
    private val mSocket = SocketHandler.getSocket()

    fun listenToAnswerEvents() {
        onFeedback()
        onBonusPoints()
        onEndGame()
        onTimesUp()
        onGradeAnswers()
        onNextQuestion()
        goToVoting()
    }

    // TODO : fix on feedback pl0x : args is empty and or null
    // server sends : Feedback :  { score: 0, answerCorrectness: 0, correctAnswer: [ '0' ] }
    fun onFeedback() {
        Log.d("answer service", "called onFeedback")
        Log.d("answer socket", "Socket is null? : id=${mSocket.id()} and ${mSocket.isActive}, and connected= ${mSocket.connected()}")
        // HAD TO MODIFY THIS CAUSE WHEN WE SEND THE FEEDBACK FROM THE SERVER TO THE HOST, THE ARGS ARE EMPTY AND FEEDBACK IS NEVER TRUE.
        mSocket.on(AnswerEvents.FEEDBACK.value) { args ->
            Log.d("answer socket", " args is empty : ${args.isEmpty()}")
            if(MatchContextService.getContext() === MatchContext.HOSTVIEW){
                showFeedback = true
            }
            if (args.isNotEmpty() && args[0] != null) {
                val jsonObject = JSONObject(args[0].toString())

                val answerCorrectnessValue = jsonObject.optInt("answerCorrectness")
                val mappedCorrectness =
                    AnswerCorrectness.entries.find { it.value == answerCorrectnessValue }
                        ?: AnswerCorrectness.WRONG

                jsonObject.put("answerCorrectness", mappedCorrectness.name)

                feedback = Gson().fromJson(jsonObject.toString(), Feedback::class.java)

                showFeedback = true
                isNextQuestionButtonEnabled = true
                processFeedback(feedback)
            } else {
                isNextQuestionButtonEnabled = true
            }
        }
    }


    fun onBonusPoints() {
        mSocket.on(AnswerEvents.BONUS.value) { args ->
            if (args.isNotEmpty() && args[0] != null) {
                bonusPoints = args[0].toString().toInt()
            }
        }
    }

    fun onTimesUp() {
        mSocket.on(AnswerEvents.TIMES_UP.value) {
            isTimesUp = true
            this.isSelectionEnabled = false
        }
    }

    fun onEndGame() {
        mSocket.on(AnswerEvents.END_GAME.value) {
            isEndGame = true
        }
    }

    fun onGradeAnswers() {
        mSocket.on(AnswerEvents.GRADE_ANSWERS.value) { args ->
            if (args.isNotEmpty() && args[0] != null) {
                gradeAnswers = true
                playersAnswers = Gson().fromJson(
                    args[0].toString(),
                    object : TypeToken<List<LongAnswerInfo>>() {}.type
                )
            }
        }
    }

    fun onNextQuestion() {
        mSocket.on(MatchEvents.GO_TO_NEXT_QUESTION.value) {
            resetStateForNewQuestion()
        }
    }

    fun resetStateForNewQuestion() {
        feedback = Feedback(0, AnswerCorrectness.WRONG)
        correctAnswer = emptyList()
        gradeAnswers = false
        isGradingComplete = false
        showFeedback = false
        isSelectionEnabled = true
        answerCorrectness = AnswerCorrectness.WRONG
        bonusPoints = 0
        isNextQuestionButtonEnabled = false
        isTimesUp = false
        isEndGame = false
        currentLongAnswer = ""
        TimeService.isTimerPaused = false
    }

    fun sendGrades() {
        gradeAnswers = false
        val gradesInfo =
            GradesInfo(matchRoomCode = MatchRoomService.getRoomCode(), grades = playersAnswers)
        println(gradesInfo)
        val gradesInfoStringified = Gson().toJson(gradesInfo)
        val gradesInfoJsonObject = JSONObject(gradesInfoStringified)
        mSocket.emit(AnswerEvents.GRADES.value, gradesInfoJsonObject)
    }

    fun handleGrading() {
        isGradingComplete = playersAnswers.all { it.score != null }
        // println("isGradingComplete$isGradingComplete")
    }

    fun selectChoice(choice: String, userInfo: UserInfo) {
        val choiceInfo = ChoiceInfo(choice, userInfo)
        val choiceInfoStringified = Gson().toJson(choiceInfo)
        val choiceInfoJsonObject = JSONObject(choiceInfoStringified)
        mSocket.emit(AnswerEvents.SELECT_CHOICE.value, choiceInfoJsonObject)
    }

    fun goToVoting(){
        mSocket.on(AnswerEvents.END_GAME.value){
            if(!MatchRoomService.isCooldown && showFeedback && MatchRoomService.isCheaterMode ){
                MatchRoomService.startedVote = true;
//                Handler().postDelayed({
//                    MatchRoomService.voteOnCheater()
//                }, 2000)
                Timer().schedule(2000) {
                    MatchRoomService.voteOnCheater()
                }
               // MatchRoomService.voteOnCheater();
            }
            MatchRoomService.startedVote = false;
        }
    }

    fun deselectChoice(choice: String, userInfo: UserInfo) {
        val choiceInfo = ChoiceInfo(choice, userInfo)
        val choiceInfoStringified = Gson().toJson(choiceInfo)
        val choiceInfoJsonObject = JSONObject(choiceInfoStringified)
        mSocket.emit(AnswerEvents.DESELECT_CHOICE.value, choiceInfoJsonObject)
    }

    fun submitAnswer(userInfo: UserInfo) {
        isSelectionEnabled = false
        val userInfoStringified = Gson().toJson(userInfo)
        val userInfoJsonObject = JSONObject(userInfoStringified)
        mSocket.emit(AnswerEvents.SUBMIT_ANSWER.value, userInfoJsonObject)
    }

    fun updateLongAnswer() {
        println("updating long answer")
        if (!isSelectionEnabled) return
        val userInfo =
            UserInfo(userId = MatchRoomService.userId, roomCode = MatchRoomService.getRoomCode())
        println(userInfo)
        val choiceInfo = ChoiceInfo(currentLongAnswer, userInfo)
        val choiceInfoStringified = Gson().toJson(choiceInfo)
        val choiceInfoJsonObject = JSONObject(choiceInfoStringified)
        mSocket.emit(AnswerEvents.UPDATE_LONG_ANSWER.value, choiceInfoJsonObject)
    }

    private fun processFeedback(feedback: Feedback) {
        feedback.correctAnswer?.let {
            correctAnswer = it
        }

        isSelectionEnabled = false
        answerCorrectness = feedback.answerCorrectness
        playerScore = feedback.score

        MatchRoomService.sendPlayersData(MatchRoomService.getRoomCode())
    }
}

File: client/src/app/services/answer/answer.service.ts
import { Injectable } from '@angular/core';
import { MatchContext } from '@app/constants/states';
import { MatchContextService } from '@app/services/match-context/match-context.service';
import { MatchRoomService } from '@app/services/match-room/match-room.service';
import { SocketHandlerService } from '@app/services/socket-handler/socket-handler.service';
import { TimeService } from '@app/services/time/time.service';
import { AnswerCorrectness } from '@common/constants/answer-correctness';
import { AnswerEvents } from '@common/events/answer.events';
import { MatchEvents } from '@common/events/match.events';
import { ChoiceInfo } from '@common/interfaces/choice-info';
import { Feedback } from '@common/interfaces/feedback';
import { GradesInfo } from '@common/interfaces/grades-info';
import { LongAnswerInfo } from '@common/interfaces/long-answer-info';
import { UserInfo } from '@common/interfaces/user-info';
import { BehaviorSubject } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class AnswerService {
    playersAnswers: LongAnswerInfo[];
    feedback: Feedback;
    gradeAnswers: boolean;
    isGradingComplete: boolean;
    isNextQuestionButtonEnabled: boolean;
    isSelectionEnabled: boolean;
    correctAnswer: string[];
    answerCorrectness: AnswerCorrectness;
    playerScore: number;
    bonusPoints: number;
    isTimesUp: boolean;
    isEndGame: boolean;
    currentLongAnswer: string;

    private showingFeedback: boolean = false;

    private showingFeedbackSubject = new BehaviorSubject<boolean>(this.showingFeedback);
    showingFeedback$ = this.showingFeedbackSubject.asObservable();

    // Allow more constructor parameters to decouple services
    // eslint-disable-next-line max-params
    constructor(
        public socketService: SocketHandlerService,
        public matchRoomService: MatchRoomService,
        private readonly matchContextService: MatchContextService,
        private readonly timeService: TimeService,
    ) {
        this.listenToAnswerEvents();
    }

    get showFeedback(): boolean {
        return this.showingFeedback;
    }

    set showFeedback(value: boolean) {
        this.showingFeedback = value;
        this.showingFeedbackSubject.next(value);
    }

    listenToAnswerEvents() {
        this.onFeedback();
        this.onBonusPoints();
        this.onEndGame();
        this.onTimesUp();
        this.onGradeAnswers();
        this.onNextQuestion();
    }

    resetStateForNewQuestion() {
        this.feedback = {} as Feedback;
        this.correctAnswer = [];
        this.gradeAnswers = false;
        this.isGradingComplete = false;
        this.showFeedback = false;
        this.isSelectionEnabled = true;
        this.answerCorrectness = AnswerCorrectness.WRONG;
        this.bonusPoints = 0;
        this.isNextQuestionButtonEnabled = false;
        this.isTimesUp = false;
        this.isEndGame = false;
        this.currentLongAnswer = '';
        this.timeService.isPanicModeDisabled = false;
        this.timeService.isTimerPaused = false;
    }


    selectChoice(choice: string, userInfo: UserInfo) {
        const choiceInfo: ChoiceInfo = { choice, userInfo };
        this.socketService.send(AnswerEvents.SelectChoice, choiceInfo);
    }

    deselectChoice(choice: string, userInfo: UserInfo) {
        const choiceInfo: ChoiceInfo = { choice, userInfo };
        this.socketService.send(AnswerEvents.DeselectChoice, choiceInfo);
    }

    submitAnswer(userInfo: UserInfo) {
        this.isSelectionEnabled = false;
        this.socketService.send(AnswerEvents.SubmitAnswer, userInfo);
    }

    updateLongAnswer() {
        if (!this.isSelectionEnabled) return;
        const userInfo = { userId: this.matchRoomService.getUserId(), roomCode: this.matchRoomService.getRoomCode() };
        const choiceInfo: ChoiceInfo = { choice: this.currentLongAnswer, userInfo };
        this.socketService.send(AnswerEvents.UpdateLongAnswer, choiceInfo);
    }

    onFeedback() {
        this.socketService.on(AnswerEvents.Feedback, (feedback: Feedback) => {
            this.feedback = feedback;
            this.showFeedback = true;
            this.isNextQuestionButtonEnabled = true;

            if (feedback) this.processFeedback(feedback);
        });
    }

    onBonusPoints() {
        this.socketService.on(AnswerEvents.Bonus, (bonus: number) => {
            this.bonusPoints = bonus;
        });
    }

    cheaterGetsBonus() {
        const totalVotes = this.matchRoomService.totalVotes.reduce((total, voteData) => total + voteData.numberOfVotes, 0);
        if (this.matchRoomService.cheaterPlayer.username === this.matchRoomService.votesData.username) {
            if (this.matchRoomService.votesData.numberOfVotes / totalVotes <= 0.5) {
                return true;
            }
        }
        return false;
    }

    onCheaterFinalScores() {
        this.socketService.on(MatchEvents.VoteOnCheater, (bonus: number) => {
            this.bonusPoints += this.bonusPoints * 0.3;
        });
    }

    onEndGame() {
        this.socketService.on(AnswerEvents.EndGame, () => {
            this.isEndGame = true;
        });
    }

    onGradeAnswers() {
        this.socketService.on(AnswerEvents.GradeAnswers, (answers: LongAnswerInfo[]) => {
            this.playersAnswers = answers;
            this.gradeAnswers = true;
        });
    }

    onNextQuestion() {
        this.socketService.on(MatchEvents.GoToNextQuestion, () => {
            this.resetStateForNewQuestion();
        });
    }

    onTimesUp() {
        this.socketService.on(AnswerEvents.TimesUp, () => {
            this.isTimesUp = true;
            this.isSelectionEnabled = false;
        });
    }

    sendGrades() {
        this.gradeAnswers = false;
        const gradesInfo: GradesInfo = { matchRoomCode: this.matchRoomService.getRoomCode(), grades: this.playersAnswers };
        this.socketService.send(AnswerEvents.Grades, gradesInfo);
    }

    handleGrading(): void {
        this.isGradingComplete = this.playersAnswers.every((answer: LongAnswerInfo) => answer.score !== null);
    }

    private processFeedback(feedback: Feedback) {
        if (this.feedback.correctAnswer) this.correctAnswer = this.feedback.correctAnswer;
        this.isSelectionEnabled = false;
        this.answerCorrectness = feedback.answerCorrectness;
        this.playerScore = feedback.score;
        this.matchRoomService.sendPlayersData(this.matchRoomService.getRoomCode());
        this.finaliseRound();
    }

    private finaliseRound() {
        const context = this.matchContextService.getContext();
        if (context === MatchContext.TestPage || context === MatchContext.RandomMode) {
            this.matchRoomService.goToNextQuestion();
            this.isNextQuestionButtonEnabled = false;
        }
    }
}
